hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -796,6 +847,307 @@ def as_result(self) -> "ConjectureResult":
+        min_value: Optional[int] = None,
+        max_value: Optional[int] = None,
+        # weights are for choosing an element index from a bounded range
+        weights: Optional[Sequence[float]] = None,
+        shrink_towards: int = 0,
+    ) -> int:
+        # This is easy to build on top of our existing conjecture utils,
+        # and it's easy to build sampled_from and weighted_coin on this.
+        if weights is not None:
+            sampler = Sampler(weights)
+            idx = sampler.sample(self._cd)
+            assert shrink_towards <= min_value  # FIXME: reorder for good shrinking
+            return range(min_value, max_value + 1)[idx]
+
+        if min_value is None and max_value is None:
+            return self._cd.unbounded_integers()
+
+        if min_value is None:
+            if max_value <= shrink_towards:
+                return max_value - abs(self._cd.unbounded_integers())
+            else:
+                probe = max_value + 1
+                while max_value < probe:
+                    self._cd.start_example(ONE_BOUND_INTEGERS_LABEL)
+                    probe = self._cd.unbounded_integers()
+                    self._cd.stop_example(discard=max_value < probe)
+                return probe
+
+        if max_value is None:
+            if min_value >= shrink_towards:
+                return min_value + abs(self._cd.unbounded_integers())
+            else:
+                probe = min_value - 1
+                while probe < min_value:
+                    self._cd.start_example(ONE_BOUND_INTEGERS_LABEL)
+                    probe = self._cd.unbounded_integers()
+                    self._cd.stop_example(discard=probe < min_value)
+                return probe
+
+        # For bounded integers, make the bounds and near-bounds more likely.
+        forced = None
+        if max_value - min_value > 127:
+            bits = self._cd.draw_bits(7)
+            forced = {
+                122: min_value,
+                123: min_value,
+                124: max_value,
+                125: max_value,
+                126: min_value + 1,
+                127: max_value - 1,
+            }.get(bits)
+
+        return self._cd.integer_range(
+            min_value,
+            max_value,
+            center=shrink_towards,
+            forced=forced,
+        )
+
+    def draw_float(
+        self,
+        *,
+        min_value: Optional[float] = None,
+        max_value: Optional[float] = None,
+        allow_nan: bool = True,
+        smallest_nonzero_magnitude: Optional[float] = None,
+        # TODO: consider supporting these float widths at the IR level in the
+        # future.
+        # width: Literal[16, 32, 64] = 64,
+        # exclude_min and exclude_max handled higher up
+    ) -> float:
+        # TODO: this does a decent bit of initialization logic *on every draw*,
+        # which is probably going to be expensive.
+        # Previously, this cost was paid per-strategy init on StringsStrategy.
```

Comment:
as noted by this comment, moving logic from `FloatStrategy.__init__` to here causes it to be invoked *per-float draw* instead of per-strategy instantiation.

Is there a way to avoid this? Caching these computations by `key = hash((min_value, max_value, allow_nan, smallest_nonzero_magnitude))` is the nuclear option, I suppose; but this becomes a question of what the cache size should be, since we can't rely on strategy lifetime to clean up the cache anymore.

Disclaimer: I have not profiled these changes.

Reply:
you weren't kidding about annoying to implement! Gotta love `hash(-0.0) == hash(0.0) and -0.0 == 0.0`.

It won't win any awards for code beauty...but it should work. We'll see how many issues you have with my approach.